pandas/_libs/tslibs/offsets.pyx

```diff
@@ -323,11 +324,17 @@ cdef _determine_offset(kwds):
 
     _kwds_use_relativedelta = ('years', 'months', 'weeks', 'days',
                                'year', 'month', 'week', 'day', 'weekday',
-                               'hour', 'minute', 'second', 'microsecond')
+                               'hour', 'minute', 'second', 'microsecond',
+                               'millisecond')
 
     use_relativedelta = False
     if len(kwds_no_nanos) > 0:
         if any(k in _kwds_use_relativedelta for k in kwds_no_nanos):
+            if "millisecond" in kwds_no_nanos:
+                raise NotImplementedError(
+                    "Using DateOffset to replace `millisecond` component in "
+                    "datetime object is not supported."
```

Comment:
Maybe a "Use X instead"?

Reply:
I'm not sure if an alternative exists here. Please advise?

Reply:
Probably `microsecond=ms*1000`?

Reply:
I find that the `microsecond` component in pandas consists of everything that's below `second` to above `nanosecond` level:
```python
In [7]: ts
Out[7]: Timestamp('2000-01-08 08:00:00.123456789')
In [14]: ts.microsecond
Out[14]: 123456
```
So to replace just the `millisecond` component, I think we need to combine the sub-millisecond fraction of the timestamp with the component to replace, e.g.:
```python
In [20]: millisecond=888
In [21]: ts + pd.DateOffset(microsecond=ts.microsecond % 1000 + millisecond * 1000)
/opt/conda/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3457: UserWarning: Discarding nonzero nanoseconds in conversion.
  exec(code_obj, self.user_global_ns, self.user_ns)
Out[21]: Timestamp('2000-01-08 08:00:00.888456789')
```
I think the it should be `timestamp.microsecond % 1000 + ms * 1000`.